designs/plugins.md

```diff
@@ -0,0 +1,255 @@
+---
+state: draft
+---
+
+
+# Node-RED Plugins
+
+## Summary
+
+We want to make it easier to extend Node-RED through the use of plugins.
+
+A plugin is an npm module that is loaded by Node-RED and used to provide custom
+functionality to either the runtime and/or editor.
+
+
+## Authors
+
+ - Nick O'Leary
+
+## Details
+
+A lot can already be done using the model we have for nodes; they are installed
+as npm modules, loaded into the runtime and the editor. The runtime manages
+their lifecycle.
+
+Whilst this works for getting custom JavaScript loaded into the editor, we want
+to have a bit more structure in place to help plugins do the right thing.
+
+The main points this design needs to cover are:
+
+ - An editor API that plugins use to register themselves - similar to
+   `RED.nodes.registerType` used by nodes. This API would then let the editor
+   manage the plugin lifecycle
+
+ - We know there are cases like nrlint where the core linter plugin will want
+   to load its own plugins.
+
+ - Loading additional resources - a plugin might not want to provide all of its
+   content in the single .html file. We've long thought about having a way for
+   nodes to declare additional resources that it will want to load in the editor.
+   The same is true for plugins.
+
+ - If a plugin is runtime-only, it should not be required to provide a .js file
+
+ - **Question:** should it be possible to dynamically remove/disable a plugin as you can nodes? That would
+   require *all* UI customisations to be undoable.
+   If plugins are there in order to provide customisations of the editor, it may not be practical to
+   fully support removing them without a reload of the editor.
+   This will require more work to figure out whilst implementing some real plugins.
+
+
+### Plugin registration
+
+The following applies to both editor and runtime plugins.
+
+```
+RED.plugins.registerPlugin("plugin-identifier", {
+    type: "plugin-type",
+    onadd: function() {},
+    onremove: function() {}
+
+})
+```
+
```

Comment:
Nrlint use `settings.js` to expose its settings to the editor using `settings`([Exposing settings to the editor](https://nodered.org/docs/creating-nodes/node-js#exposing-settings-to-the-editor) for Custom node).
This may be useful for another plugins.

Reply:
Yes - that is a good suggestion. I will look at including that here.

Reply:
With the design (and draft implementation), a plugin can be provided as just the HTML file. The problem is then there is nothing loaded in the runtime that includes the settings configuration to tell the runtime what settings can/should be exported to the editor. So does that mean every plugin that wants to expose settings to the editor must have a .js file as well to *only* register the settings to export.

It would be nice not to have to do that for editor-only plugins.

The naming scheme used by node settings is also quite awkward and this might be an opportunity to rethink how it works and make it better.

Still thinking about this....